charlotte_core/src/memory/pmm.rs

```diff
@@ -0,0 +1,243 @@
+    Mutex::new(
+        bootinfo::HHDM_REQUEST
+            .get_response()
+            .expect("Limine failed to create a direct mapping of physical memory.")
+            .offset() as PhysicalAddress,
+    )
+});
+
+pub static PHYSICAL_FRAME_ALLOCATOR: Lazy<Mutex<PhysicalFrameAllocator>> = Lazy::new(|| {
+    Mutex::new(PhysicalFrameAllocator::new())
+});
+
+
+type PhysicalAddress = usize;
+
+pub struct MemoryMap {
+    entries: &'static [&'static bootinfo::memory_map::Entry],
+}
+
+impl MemoryMap {
+    pub fn get() -> MemoryMap {
+        MemoryMap {
+            entries: bootinfo::MEMORY_MAP_REQUEST
+                .get_response()
+                .expect("Limine failed to obtain a memory map.")
+                .entries(),
+        }
+    }
+    pub fn total_memory(&self) -> usize {
+        self.entries
+            .iter()
+            .filter(|entry| entry.entry_type != bootinfo::memory_map::EntryType::BAD_MEMORY)
+            .map(|entry| entry.length)
+            .sum::<u64>() as usize
+    }
+    pub fn usable_memory(&self) -> usize {
+        self.entries
+            .iter()
+            .filter(|entry| entry.entry_type == bootinfo::memory_map::EntryType::USABLE)
+            .map(|entry| entry.length)
+            .sum::<u64>() as usize
+    }
+    pub fn iter(&self) -> core::slice::Iter<&bootinfo::memory_map::Entry> {
+        self.entries.iter()
+    }
+    pub fn find_best_fit(&self, size: usize) -> Result<&bootinfo::memory_map::Entry, Error> {
+        self.entries
+            .iter()
+            .filter(|entry| entry.entry_type == bootinfo::memory_map::EntryType::USABLE)
+            .filter(|entry| entry.length as usize >= size)
+            .min_by_key(|entry| entry.length)
+            .ok_or(Error::InsufficientContiguousMemoryAvailable)
+            .copied()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Error {
+    OutOfMemory,
+    InsufficientMemoryAvailable,
+    InsufficientContiguousMemoryAvailable,
+    MemoryOvercommitted,
+    AddressMisaligned,
+    AddressOutOfRange,
+    InvalidSize,
+    InvalidAlignment,
+}
+
+enum RegionAvailability {
+    Available,
+    Unavailable(PhysicalAddress),
+}
+
+const FRAME_SIZE: usize = 4096;
+
+/// A bitmap based physical frame allocator
+pub struct PhysicalFrameAllocator {
+    bitmap: &'static mut [u8]
+}
+
+impl PhysicalFrameAllocator {
+    fn new() -> PhysicalFrameAllocator {
+        let memory_map = MemoryMap::get();
+        let total_memory = &memory_map.total_memory();
+        // find a region that is large enough to hold the bitmap
+        let region = &memory_map.find_best_fit(total_memory / FRAME_SIZE / 8)
+            .expect("Failed to find a physical memory region large enough to hold the physical frame allocator bitmap");
+        // create the PFA
+        let pfa = PhysicalFrameAllocator {
+            // Safety: The region is guaranteed to be valid and the length is guaranteed to be large enough to hold the bitmap
```

Comment:
Remark: I like this idea of putting safety comments on unsafe blocks, we should probably do this more often.

Reply:
Since unsafe is all about doing things where the compiler can't prove all the needed invariants for us, it makes sense to make note of the fact that those invariants do still hold and how we know that. This comment should probably actually be a bit more informative on that.

I'll add more to it on the next refactor.